Observer refactoring project

szegedim opened the pull request:
This is our candidate to be merged into master first. It ensures that any observer changes we have runs like the one in master. It checks whether the observer localnet container runs with the right settings. +@pradeep Palanisamy the other author.
Once it is merged, we plan to give you bite sized pull requests every day. This should not overwhelm you. It will help to catch up with our branches quickly.

Authors: Miklos Szegedi, Pradeep Palanisamy

smnzhu reviewed (commented):
why did we define a new root `observer` directory?

It seems like the files under that directory should be placed under `cmd/access` or `cmd/observer`?

smnzhu reviewed (commented):
To reiterate, this is the desired directory structure we want after the refactoring:

* `cmd/access` which contains all the builder code for Access Node
* `cmd/observer` which contains all the builder code for the Observer Node

You should look at the existing node builder code under `cmd` (e.g. `cmd/consensus`, `cmd/execution`) to see examples of what we want.

Currently, `staked_access_node_builder.go` and `unstaked_access_node_builder.go` contain a lot of shared code, which is in `access_node_builder.go`. After the refactor, we want to completely separate these, and the builder code under `cmd/access` and `cmd/observer` should *not* contain any shared code (except to the extent that they both use `FlowNodeBuilder`)

smnzhu commented on cmd/access/node_builder/staked_access_node_builder.go:
```diff
@@ -221,6 +221,7 @@ func (builder *StakedAccessNodeBuilder) Build() (cmd.Node, error) {
 				builder.rpcMetricsEnabled,
 				builder.apiRatelimits,
 				builder.apiBurstlimits,
+				nil,
```
this file should no longer exist after the refactor.

szegedim replied:
I planned to add an changelist once the concepts and earlier changelists are approved.

szegedim replied:
This is the final layout: https://github.com/GetElastech/flow-go/commit/33623bb268ecfaedd1680372860ab98e6a35b7b6
```
cmd/access/
cmd/access/main.go
cmd/access/README.md
cmd/access/.gitignore
cmd/access/node_builder
cmd/access/node_builder/access_node_builder.go```

smnzhu commented on cmd/access/node_builder/unstaked_access_node_builder.go:
```diff
@@ -312,6 +313,30 @@ func (builder *UnstakedAccessNodeBuilder) enqueueUnstakedNetworkInit() {
 func (builder *UnstakedAccessNodeBuilder) enqueueConnectWithStakedAN() {
 	builder.Component("upstream connector", func(_ *cmd.NodeConfig) (module.ReadyDoneAware, error) {
 		return newUpstreamConnector(builder.bootstrapIdentities, builder.LibP2PNode, builder.Logger), nil
+	}).Component("RPC engine", func(node *cmd.NodeConfig) (module.ReadyDoneAware, error) {
```
this file should no longer exist after the refactor.

szegedim replied:
Indeed. I will add a changelist.

szegedim replied:
This is the final layout: https://github.com/GetElastech/flow-go/commit/33623bb268ecfaedd1680372860ab98e6a35b7b6
```
cmd/access/
cmd/access/main.go
cmd/access/README.md
cmd/access/.gitignore
cmd/access/node_builder
cmd/access/node_builder/access_node_builder.go```

szegedim commented:
@smnzhu "why did we define a new root observer directory?"
This makes the code reusable (e.g. follower/) The reason is that we have now a service that can be used from the follower library.
cmd/observer is reserved to the cmd, which is the endpoint of an application

smnzhu commented:
> @smnzhu "why did we define a new root observer directory?" This makes the code reusable (e.g. follower/) The reason is that we have now a service that can be used from the follower library. cmd/observer is reserved to the cmd, which is the endpoint of an application

Hmm, but the `observer` directory seems to contain a lot of stuff that would not be relevant to the follower library (e.g. command line flag parsing). The only thing that the Consensus Follower library really needs is the Sync Engine and the Follower Engine (and any dependencies of those two), it doesn't need any other modules / components that are used by the Observer. So I think it's still better to completely separate the Access Node, the Observer, and the Consensus Follower. The amount of code duplication between the Observer and the Consensus Follower will be minimal.

szegedim commented:
It seems that a change got lost during the rebase from feature/observer-prc-0509d. Fixing.

szegedim commented:
@smnzhu This is the final status of the follower library. I simplified all three.
```
follower/
follower/consensus_follower.go
follower/follower_builder.go
```

szegedim commented:
Replaced by: https://github.com/onflow/flow-go/pull/2554
It addresses the issues above, creating separate simple access, observer and follower commands, and moving `upstream` into `module/` as requested by @smnzhu above.
